### PR TITLE
Avoiding using un-set repeated values from proto.

### LINKED ISSUE #138: There is a difference between `[]` and `None`
I have a Model:
```
class foo(EndpointsModel):
    bar = NDB.KeyProperty(kind='baz', repeated=True)

@foo.method(...)
def update(self, f):
   f.put()
   return f
```

when I use the API explorer and enter `{ bar: [...]}`everything works and `bar` is set correctly.

but when I do not set `bar` in the request , `bar` is implicitly set to `[]` and my previous data is lost.

I would like the data to be lost only when I explicitly set `{bar: []}`

### PR DESCRIPTION
**Pull request by dhermes**
Fixes #138.

See https://gist.github.com/dhermes/29514e4926bef1837ebd for a test app and the differences before this PR and after it.

/cc @matantsu @kdeus

### CODE REVIEW

**Comment by kdeus** [comment]
How's this work if the user explicitly sets {bar: []}?  Isn't bar left alone, rather than cleared, with this change?

**Comment by dhermes** [comment]
Thanks for making me check myself before I wreck myself @kdeus. Luckily I have an automated test app so it'll be easier to check the behavior!

**Comment by dhermes** [comment]
@kdeus It seems it would be easiest to just keep the keys when the message is decoded:

```python
import json

from endpoints import protojson


class DJHProtoJson(protojson.EndpointsProtoJson):

    def decode_message(self, message_type, encoded_message):
        if not encoded_message.strip():
            return message_type()

        dictionary = json.loads(encoded_message)
        message = self._ProtoJson__decode_dictionary(message_type, dictionary)
        message.check_initialized()
        # Need to begin with _Message to make the Message metaclass happy.
        message._Message__decoded_fields = dictionary.keys()
        return message
```

Any suggestions for how to register a custom `protojson` class from within `endpoints_proto_datastore`?

**Comment by kdeus** [comment]
Endpoints does this already...  You can probably do something similar, though I'm not sure if that'll end up clashing with our replacement:

https://chromium.googlesource.com/external/googleappengine/python/+/c10d3bdc6af30cb71cfb8a5a8551450e8ccce29c/google/appengine/ext/endpoints/apiserving.py#238

Alternatively, you could try to hack our replacement:

https://chromium.googlesource.com/external/googleappengine/python/+/c10d3bdc6af30cb71cfb8a5a8551450e8ccce29c/google/appengine/ext/endpoints/protojson.py

**Comment by dhermes** [comment]
Will do. Thanks for the links. I'll ping you once I "figure it out".

**Comment by dhermes** [comment]
@kdeus PTAL. Note the update to `output_after.txt` in the gist (i.e. it works as expected).

File: endpoints_proto_datastore/utils.py [hunk 1]
```diff
@@ -417,3 +419,40 @@ def query_method(modelclass, **kwargs):
         query method.
   """
   return _GetEndpointsMethodDecorator('query_method', modelclass, **kwargs)
+
+
+class _EPDProtoJson(protojson.EndpointsProtoJson):
+  """Slightly modifed version of EndpointsProtoJson.
+
+  The key difference is that when parsing a message, the dictionary keys
+  or stored on the newly created message.
```
Comment (kdeus) [inline endpoints_proto_datastore/utils.py]: or -> are

File: endpoints_proto_datastore/utils.py [hunk 2]
```diff
@@ -417,3 +419,40 @@ def query_method(modelclass, **kwargs):
...
+
+    Raises:
+      ValueError: If encoded_message is not valid JSON.
+      messages.ValidationError if merged message is not initialized.
+    """
+    if not encoded_message.strip():
+      return message_type()
+
+    msg_dictionary = json.loads(encoded_message)
+    message = self._ProtoJson__decode_dictionary(message_type, msg_dictionary)
```
Comment (kdeus) [inline endpoints_proto_datastore/utils.py]: Why not use super() here?  If EndpointsProtoJson ever adds a decode_message method, this is going to skip it...

Comment (dhermes) [inline endpoints_proto_datastore/utils.py]: I'm like 99.99% certain this isn't possible with "private" (i.e. starting with `__`) method names. I don't actually know why anyone would ever use `__`, it's lame and doesn't provide anything other than obfuscation.

File: endpoints_proto_datastore/utils.py (hunk 1 above)
Comment (dhermes) [inline endpoints_proto_datastore/utils.py]: DERP. Thanks.

Comment (dhermes) [inline endpoints_proto_datastore/utils.py]: Fixed and folded into the original commit.

File: endpoints_proto_datastore/utils.py (hunk 2 above)
Comment (dhermes) [inline endpoints_proto_datastore/utils.py]: Yeah I played around with this and subclasses don't inherit the name `__foo`, so subclasses would actually need to implement `_ClassName__foo`, so if you over-rode it, we'd just need to use yours.

Comment (kdeus) [inline endpoints_proto_datastore/utils.py]: Alright.  It's unlikely EndpointsProtoJson will change anytime soon.  If it does, I'll keep this in mind.

Comment (dhermes) [inline endpoints_proto_datastore/utils.py]: Thanks.

**Comment by dhermes** [comment]
@kdeus Anything else to do?

**Comment by kdeus** [comment]
LGTM.

**Comment by dhermes** [comment]
Woo hoo! Thanks Kev!

**Pull request merged**